core/server/kustomization.go

```diff
@@ -40,19 +41,47 @@ func (cs *coreServer) ListKustomizations(ctx context.Context, msg *pb.ListKustom
 		return nil, defaultClusterNotFound{}
 	}
 
-	for _, ns := range nsList {
-		nsResult, err := listKustomizationsInNamespace(ctx, clustersClient, ns.Name)
-		if err != nil {
-			cs.logger.Error(err, fmt.Sprintf("unable to list kustomizations in namespace: %s", ns.Name))
+	newNextPageToken := ""
 
-			continue
+	// TODO: Once the UI handles pagination we can remove this if block.
+	// It was left like that so the UI doesn't brake with backend pagination changes
+	if msg.Pagination == nil {
+		for _, ns := range nsList {
+			nsResult, _, err := listKustomizationsInNamespace(ctx, clustersClient, ns.Name, 0, "")
+			if err != nil {
+				cs.logger.Error(err, fmt.Sprintf("unable to list kustomizations in namespace: %s", ns.Name))
+
+				continue
+			}
+
+			results = append(results, nsResult...)
+		}
+	} else {
+
+		newNextPageToken, err = GetNextPage(
+			nsList,
+			msg.Pagination.PageSize,
+			msg.Pagination.PageToken,
+			func(namespace string, limit int32, pageToken string) (string, int32, error) {
+				nsResult, nextK8sPageToken, err := listKustomizationsInNamespace(ctx, clustersClient, namespace, limit, pageToken)
```

Comment:
If I understand correctly, the behavior of this pagination is not how ppl are thinking about pagination.

If the UI wants to get 50 entries, it will expect to get 50 of them. But this will return 50 from each namespaces.

For example if we have 5 namespaces and each has 200 entries, the UI makes a request with `limit: 50`, this endpoint will return 250 (5*50) per page and not 50.

I don't have a solution to this, just wanted to take a note about it. I don't know how can we manage this, because all possible solutions has it's drawback and has to be a decision about an opinionated solution. For example,
what what counts as first N entries? Based on namespace and read until we reach N?  Can we even use the pagination API if we go down on that route?

Reply:
Interesting guess, I'm really curious in knowing what would be making you think that could happen. I'll try to clarify the logic. The key relies on `itemsLeft`. Every time we get a page from the k8s API, I decrease `itemsLeft` by the size of that page, causing it to eventually get to 0 (if there are enough resources). If there are NOT enough resources the loop will simply finish and return only the items it gathered. 

Probably the confusion relies on the name of the variables in this function. The parameter `limit` received here is not always the limit coming from `msg.Pagination.PageSize`, let's remember that this function could be called N times, it could be 1 it could be 100. It all depends on how many kustomizations and namespaces are there. It might help if I change them from `(namespace string, limit int32, pageToken string)` to `(currentQueryNamespace string, k8sLimit string, k8sPageToken string)`? or something similar. What do you think @yitsushi ? Any suggestion for the name of these parameters? 🙂

Reply:
Ok, nvm, just ignore me 😆 

I read the whole flow again very carefully and now I see what's happening. I missed the part where we create an intermediate nextToken that contains the original k8s nextToken (from `Continue`) and the index of the namespace where we left, so I though we are using the same token and that would be bound to one request on a namespace, and from that my brain drew the conclusion it's call on all namespaces with the provided limit

Reply:
No worries, I'm glad we are on the same page, thanks for the involvement though, I appreciate it. 🙂